src/typechecker.erl

```diff
@@ -4308,11 +4302,32 @@ add_type_pat(OpPat = {op, _Anno, _Op, _Pat}, Ty, Env, VEnv) ->
 add_type_pat(Pat, Ty, _Env, _VEnv) ->
     throw({type_error, pattern, element(2, Pat), Pat, Ty}).
 
+-spec expect_map_type(type(), env()) -> R when
+      R :: any
+         | {assoc_tys, [type()] | any, constraints:constraints()}
+         | {type_error, type()}.
+expect_map_type(?type(any), _Env) ->
+    any;
+expect_map_type(?top(), _Env) ->
+    {assoc_tys, [type(map_field_assoc, [top(), top()])], constraints:empty()};
+expect_map_type({var, _, Var}, _Env) ->
+    %% FIXME this is a quite rudimentary implementation
+    %% - variables from the map pattern become any()
+    %% - the constraint could contain the map keys
+    Cs = constraints:add_var(Var, constraints:upper(Var, type(map, any))),
+    {assoc_tys, any, Cs};
+expect_map_type(?type(map, AssocTys), _Env) ->
+    {assoc_tys, AssocTys, constraints:empty()}.
+
 %% Rewrite map_field_assoc to map_field_exact to return in pattern types.
 %%
 %% Similarly to map field type inference on map creation - if a pattern matches,
 %% then the map field is exact (:=), not assoc (=>).
 %% There isn't even syntax for optional fields in map patterns.
+rewrite_map_assocs_to_exacts(?top()) ->
+    top();
+rewrite_map_assocs_to_exacts({var, _, _Var}) ->
+    type(none);
```

Comment:
What are these clauses doing here? Var :: none() seems very weird to me.

I though we only call this function on explicit map types.

Reply:
That was the case before introducing `expect_map_type`. Now multiple of the former `add_type_pat` clauses fall into case clause for `{assoc_tys, ...}`, therefore all their return values have to be handled here, too, unless we want to explicitly branch there 🤷‍♂️ . That's just a trade-off between where to branch, nothing new introduced here.

Reply:
Ahh, and it's suggested when it can be `none` here:
https://github.com/josefs/Gradualizer/blob/c3ee078616a88fc25655cb4bc40e81cda2f01cd3/src/typechecker.erl#L4036

Reply:
Yeah there are some strange types used for refinement. If we spill that over to other functions, we'd have to describe it there too... Better keep the semantics of each function simple, so maybe branch before this function is called then..?

Reply:
Cleaned up now :+1: